src/lib.rs

```diff
@@ -0,0 +1,989 @@
+pub mod shape;
+pub mod iter;
+
+use num::Num;
+use std::ops::Add;
+use std::ops::Sub;
+use std::ops::Mul;
+use std::ops::Div;
+
+use crate::shape::Shape;
+use crate::iter::IndexIterator;
+
+pub type Axes = Vec<usize>;
+
+#[derive(Debug, Clone)]
+pub struct Tensor<T: Num> {
```

Comment:
I think it would be best if we had an `AbstractTensor` trait that `Tensor` implements since we'll have various different types of tensors (e.g. `SparseTensor`, `StaticTensor`).

Reply:
Yeah I agree. Was thinking of calling it `RawTensor`. That might be semantically slightly different though.

Reply:
That sounds a lot like inheritance - which is not what traits are for. Rather we should have SparseTensors, StaticTensors and DynamicTensors which are composed of reusable parts (like Shape _maybe_).

Reply:
We probably will want a trait which outlines the basic operations for a generic tensor.